Readme.md

```diff
@@ -0,0 +1,375 @@
+
+Note over CA,CRA: Job with id JobId for Client A was started earlier
+CA->>S: Status(JobId)
+S->>CRA: Status for JobId pls
+alt Job still running
+CRA->>S: Id exists, no status, must be running
+S->>CA: running
+else Job exited
+CRA->>S: Id exists, has exit_code
+S->>CA: exit_code
+else Client kills Job
+CRA->>S: Id exists, has signal
+S->>CA: signal
+end 
+```
+</details>
+
+### Output
+
+[![](https://mermaid.ink/img/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG5hY3RvciBDQSBhcyBDbGllbnQgQVxucGFydGljaXBhbnQgUyBhcyBKb2IgUnVubmVyIFNlcnZpY2UgXG5wYXJ0aWNpcGFudCBDUkEgYXMgQ29udHJvbGxlciBmb3IgQ2xpZW50IEFcbnBhcnRpY2lwYW50IEogYXMgSm9iXG5wYXJ0aWNpcGFudCBGUyBhcyBGaWxlIFN5c3RlbVxuXG5Ob3RlIG92ZXIgQ0EsRlM6IEpvYiB3aXRoIGlkIEpvYklkIGZvciBDbGllbnQgQSB3YXMgc3RhcnRlZCBlYXJsaWVyXG5KLSlGUzogU3RyZWFtcyBsb2dzXG5DQS0-PlM6IE91dHB1dChKb2JJZClcblMtPj5DUkE6IE91dHB1dCBmb3IgSm9iSWQgcGxzXG5DUkEtPj5GUzogT3BlbiBMb2cgZmlsZXMgZm9yIEpvYklkXG5GUy0-PkNSQTogSGVyZSBpcyB0aGUgaGFuZGxlXG5DUkEtPj5TOiBTdHJlYW0gY29udGVudHMgYmFja1xuUy0-PkNBOiBIZXJlIGFyZSB5b3VyIGxvZ3MiLCJtZXJtYWlkIjp7InRoZW1lIjoiZGVmYXVsdCJ9LCJ1cGRhdGVFZGl0b3IiOmZhbHNlLCJhdXRvU3luYyI6dHJ1ZSwidXBkYXRlRGlhZ3JhbSI6ZmFsc2V9)](https://mermaid.live/edit#eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG5hY3RvciBDQSBhcyBDbGllbnQgQVxucGFydGljaXBhbnQgUyBhcyBKb2IgUnVubmVyIFNlcnZpY2UgXG5wYXJ0aWNpcGFudCBDUkEgYXMgQ29udHJvbGxlciBmb3IgQ2xpZW50IEFcbnBhcnRpY2lwYW50IEogYXMgSm9iXG5wYXJ0aWNpcGFudCBGUyBhcyBGaWxlIFN5c3RlbVxuXG5Ob3RlIG92ZXIgQ0EsRlM6IEpvYiB3aXRoIGlkIEpvYklkIGZvciBDbGllbnQgQSB3YXMgc3RhcnRlZCBlYXJsaWVyXG5KLSlGUzogU3RyZWFtcyBsb2dzXG5DQS0-PlM6IE91dHB1dChKb2JJZClcblMtPj5DUkE6IE91dHB1dCBmb3IgSm9iSWQgcGxzXG5DUkEtPj5GUzogT3BlbiBMb2cgZmlsZXMgZm9yIEpvYklkXG5GUy0-PkNSQTogSGVyZSBpcyB0aGUgaGFuZGxlXG5DUkEtPj5TOiBTdHJlYW0gY29udGVudHMgYmFja1xuUy0-PkNBOiBIZXJlIGFyZSB5b3VyIGxvZ3MiLCJtZXJtYWlkIjoie1xuICBcInRoZW1lXCI6IFwiZGVmYXVsdFwiXG59IiwidXBkYXRlRWRpdG9yIjpmYWxzZSwiYXV0b1N5bmMiOnRydWUsInVwZGF0ZURpYWdyYW0iOmZhbHNlfQ)
+
+Notes:
+- Considered keeping logs in-memory, but I believe this approach has couple of drawbacks. Reads and writes need to be synchronized somehow, so our options are either introducing Mutex, which may become fairly hot (depends on the volume of logs), or making Job own the output of the child process instead of streaming to file. This consequently means that all job output is now in main process memory, regardless of whether it will ever be read or not. 
+- Assuming job is still running, only logs that were generated at the moment of request would be sent back. To address that, controller could leverage [inotify](https://man7.org/linux/man-pages/man7/inotify.7.html) (intention is to use [inotify-rs](https://docs.rs/inotify/latest/inotify/)). Given log streamer task read EOF from open file, instead of exiting it `select!`'s on one of two events - either job done, in which case no new logs were generated or `IN_MODIFY` event, in which case file is opened again, `seek`'ed to wherever cursor was left and streaming to client resumes. 
```

Comment:
> only logs that were generated at the moment of request would be sent back

Just wanted to clarify what this means exactly - will the client still receive all logs produced by the job from the beginning? That is part of the challenge spec requirements.